lib/pinyin.ex

```diff
@@ -100,28 +100,60 @@ defmodule Pinyin do
   ## Examples
 
       iex> Pinyin.from_marked("nǐ")
-      %Pinyin{tone: 3, word: "ni"}
+      %Pinyin{tone: 3, final: "i", initial: "n"}
 
       iex> Pinyin.from_marked("nǐ")
-      %Pinyin{tone: 3, word: "ni"}
+      %Pinyin{tone: 3, final: "i", initial: "n"}
 
       iex> Pinyin.from_marked("nǐhǎo")
       ** (ArgumentError) Multiple tone marks present in 'nǐhǎo'
 
+      iex> Pinyin.from_marked("hello")
+      ** (ArgumentError) No Pinyin word found in 'hello'
+
   """
   @spec from_marked(String.t()) :: t()
   def from_marked(word) do
-    {word, tone} =
-      word
+    {:ok, parsed, _, _, _, _} = Pinyin.Parsers.pinyin_words(word)
```

Comment:
This is what I meant here:

> This remark is not entirely clear to me:
>
>> There's only the case that somebody provides the pinyin and wants to output pinyin again, where it would add an additional parse step.

Before we didn't have to parse, we just put the string into `%Pinyin`. Now we have to parse, to split it.
This is only used during compile time when parsing the `.map` file.
I didn't see a noticeable increase in compile.
I'll definitely try to measure that though.

Reply:
> The main point of the `from_marked` and `from_numbered` functions was that they could be used to "skip" the parser if you already knew the type of Pinyin you were dealing with (e.g. you could use `from_marked` to parse a string like "hǎo3") . I think we should consider dropping these functions if they don't work together with the new representation.

Makes sense, let me see if that works out.

> Down the line, I could maybe introduce an option for the pinyin parser which only considers numbered/marked characters.

Hm, makes sense, too. You mean just a faster version that avoids full parsing but is only able to convert between numbered and marked (and that only if the caller guarantees it's a valid pinyin syllable)?

Reply:
> > Down the line, I could maybe introduce an option for the pinyin parser which only considers numbered/marked characters.
>
> Hm, makes sense, too. You mean just a faster version that avoids parsing but is only able to convert between numbered and marked (and that only if the caller guarantees it's a valid pinyin syllable)?

It's not about the speed but about the guarantees the function gives you. `from_marked` and `from_numbered` were very simple, so they could handle weird cases not handled by the parser. For instance, `from_marked` would correctly parse "hao3" as hao + the number 3 while `read` would always parse it as "hao" with the third tone.

While I don't think it is very common to mix tone marked pinyin with numbers without proper spaces I think it's might be inconvenient to lose that functionality. Since the parser already has a different code path for numbered and marked pinyin I might introduce a way to use the parser and specify whether the pinyin you provide is marked or numbered.

For now, I would propose to remove `from_marked` and `from_numbered` and I'll put the parser adjustements on my TODO list. I'll see if I can ever get to it 😅.

Reply:
The new parser can parse numbered pinyin, marked pinyin, or both at the same time. Just need to make a nice frontend for it and I can add it to this PR.